client/blocks/app-banner/index.jsx

```diff
@@ -218,6 +193,45 @@ export class AppBanner extends Component {
 	}
 }
 
+export function getiOSDeepLink( currentRoute, currentSection ) {
+	const baseURI = 'https://apps.wordpress.com/get';
+	const fragment = buildDeepLinkFragment( currentRoute, currentSection );
+
+	if ( fragment.length === 0 ) {
+		return baseURI;
+	}
+
+	return `${ baseURI }#${ fragment }`;
+}
+
+export function buildDeepLinkFragment( currentRoute, currentSection ) {
+	const hasRoute = currentRoute !== null && currentRoute !== '/';
+	let fragment;
+
+	switch ( currentSection ) {
+		case EDITOR:
+			fragment = '/post';
+			break;
+		case NOTES:
+			fragment = '/notifications';
+			break;
+		case READER:
+			// The Reader is generally accessed at the root of WordPress.com ('/').
+			// In this case, we need to manually add the section name to the
+			// URL so that the iOS app knows which section to open.
+			fragment = hasRoute ? currentRoute : '/read';
+			break;
+		case STATS:
+			fragment = hasRoute ? currentRoute : '/stats';
+			break;
+		default:
+			fragment = '';
+			break;
+	}
+
+	return encodeURIComponent( fragment );
```

Comment:
just in case you are interested there are two ways we can remove the mutation here

we can return directly…

```js
const hasRoute = …

switch ( currentSection ) {
	case EDITOR:
		return encodeURIComponent( '/post' );

	case NOTES:
		return encodeURIComponent( '/notifications' );

	…

	default:
		return '';
}
```

we can also separate and compose the functionality in various ways. since this is pretty specific behavior and the function doesn't do much else, I'm going to create a function inside of it. could legitimately create another separate and testable function

```js
export function buildDeepLinkFragment( currentRoute, currentSection ) {
	const hasRoute = …;

	const getFragment = () => {
		switch ( currentSection ) {
			case EDITOR:
				return '/post';

			case NOTES:
				return '/notifications';

			case READER:
				return hasRoute ? currentRoute : '/read';

			case STATS:
				return hasRoute ? currentRoute : '/stats';

			default:
				return '';
		}
	}

	return encodeURIComponent( getFragment() );
}
```

this is mainly a style thing, just for inspiration

Reply:
Nice – I like the second option here (I wanted to avoid duplicating `encodeURIComponent` for each return).